_posts/2020-10-27-a-closer-look-at-the-interpreter.md

```diff
@@ -0,0 +1,194 @@
+     * 32 bits) and then read its contents.
+     *
+     * Note that we address the X registers directly;
+     * had this instruction not been specialized, we
+     * would first need to determine whether the
+     * argument was an X or a Y register. */
+    term = x_registers[arg_word >> 32];
+
+    /* is_not_list(term) */
+    if (term & (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)) {
+
+        /* Unpack our fail label (lower 32 bits) and add
+         * it to our current instruction pointer. This
+         * is an offset and may be negative for backward
+         * jumps. */
+        I += (Sint32)arg_word;
+
+        /* Jump to the instruction at the fail label using
+         * the "labels as values" GCC extension. */
+        goto *I;
+    }
+
+    /* Skip the current label address and argument word,
+     * then jump to the next instruction. This was
+     * automatically generated by beam_makeops. */
+    I += 2;
+    goto *I;
+}
+```
+
+There's a bit more to it, and those who would like to know more can read the
+[documentation for the `beam_makeops` script], but the above covers the gist of
+it.
+
+While the above is quite efficient, there's significant overhead from dispatch
+and argument handling. Here's a slightly altered assembly listing of the
+above:
+
+<pre class="highlight">
+    <em>; Read the argument word from the instruction
+    ; stream.</em>
+    mov    rdx, [rbx + 8]
+
+    <em>; Unpack the offset of our source register (upper 32
+    ; 32 bits).</em>
+    mov    rcx, rdx
+    shr    rcx, 32
+
+    <em>; X registers live in machine register r15, and we
+    ; placed our offset in rcx, so we can find our term at
+    ; [r15 + rcx].
+    ;
+    ; Perform a non-destructive bitwise AND on the term
+    ; using the `test` instruction, and jump to the fail
+    ; label if the result is non-zero.</em>
+    <b>test   byte [r15 + rcx], (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)
+    jne    jump_to_fail_label</b>
+
+    <em>; Skip the current label address and argument word,
+    ; then jump to the next instruction.</em>
+    add    rbx, 16
+    jmp    [rbx]
+
+jump_to_fail_label:
+    <em>; Unpack our fail label (lower 32 bits) and add
+    ; it to our current instruction pointer.</em>
+    movsxd rdx, edx
+    lea    rbx, [rbx + rdx * 8]
+
+    <em>; Jump to the instruction at the fail label.</em>
+    jmp    [rbx]
+</pre>
+
+The bold section is the meat of the instruction and the rest is argument
+unpacking and instruction dispatch. While this is not much of a problem for
+large instructions, its effect on short ones like this is very large, and when
+looking through a profiler (e.g. [`perf`]) it's not unusual for the final `jmp`
+to dominate the rest.
+
+To lessen this effect, the loader combines commonly used instruction sequences
+into a single instruction. For example, two independent moves may fuse into
+[`move2_par`], and `is_nonempty_list` followed by `get_list` might be fused to
```

Comment:
Don't use the word fuse like that. I can't even find that use in the correct meaning. Fuse together would be better , but merge or combine feels better